### ISSUE TITLE
Trailing Whitespace in MapServer EXPRESSION

### ISSUE THREAD

**Issue by glk-24** [MapServer/MapServer#6292]
## Expected behavior and actual behavior.

Trailing spaces are removed from the string attribute in the `EXPRESSION`.
So the feature isn't displayed in the returned image.
Leading spaces working fine.
The same behaviour occur with a GetFeatureInfo-Request. (leading spaces transfered while trailing spaces get trimmed).

Tested with a PostGIS layer.

e.g.

**db:**
feature | test
1 | "abc "
2 | " abc"

**mapserver mapfile:**
`EXPRESSION ('[test]' eq 'abc ')` - feature 1 not displayed
`EXPRESSION ('[test]' eq 'abc')` - feature 1 displayed
`EXPRESSION ('[test]' eq ' abc')` - feature 2 displayed
`EXPRESSION ('[test]' eq 'abc')` - feature 2 not displayed

## Operating system

CentOS Linux release 7.6.1810 (Core)

## MapServer version and installation method

7.6.2

**Comment by sdlime** [comment]
@glk-24, out of curiosity does the same thing happen with a shapefile? --Steve

**Comment by glk-24** [comment]
@sdlime Unfortunaly i can't create or edit a shapefile with trailing whitespace. Global Mapper, QGIS and even Open Office Calc trim leading and trailing whitespace on attributes when saving the dbf-file.

**Comment by jratike80** [comment]
Isn't dbf fields padded with spaces till full width? Then trailing whitespace would not make sense really.

**Comment by jmckenna** [comment]
Small test case, includes SpatialiteDB & mapfile & commands.txt, to reproduce: https://gatewaygeomatics.com/dl/ticket-6292-trailing-whitespace.zip  

Similar to @glk-24's report:
```
       EXPRESSION "Lake Superior " #map is generated
       EXPRESSION "Lake Superior"  #no map generated  

       *where the attribute value was modified to include a space in the database table, such as "Lake Superior "
```

**Comment by jmckenna** [comment]
Ah, I made it too fast to realize now that the SpatialiteDB case seems to be working as expected...  will try now with a small PostGIS test...

**Comment by jmckenna** [comment]
@glk-24 I've updated the small test case, to include a mapfile with 3 layers (PostGIS, SpatiaLite, Shapefile), and included a dump of the PostGIS 'lakes' database (5 lakes) that you can restore locally, and use with that mapfile. (the spatialite & shapefiles are also included)

To summarize: I cannot reproduce this issue on PostGIS (PostgreSQL Version: 13.2, PostGIS Version: 3.1.1), SpatiaLite, or Shapefile.  Tested on Windows and Ubuntu.

Please use that test case to test locally.  Thanks!

**Comment by glk-24** [comment]
@jmckenna Many thanks for the test case you provided. Now i'm a little confused. I've only tested the PostGIS layer, because this was my initial issue. This is what i've done:

I couldn't restore the dump in my developing system (CentOS 7.6, PostgreSQL Version: 11.2, PostGIS Version: 2.5.2) because of the version mismatch of the binary dump (pg_dump 13.2 vs. pg_restore 11.2). So i installed a fresh PostgreSQL on my local Windows machine (PostgreSQL Version: 13.2 with PostGIS Version: 3.1.1), restored the binary data and created a second dump (which i've attached: [pgdump-lakes-2021-04-27.zip](https://github.com/MapServer/MapServer/files/6382864/pgdump-lakes-2021-04-27.zip)) in plain text format, to avoid version issues.

After importing this dump in my developing system i could use your test case.
The result is the same as your comments in the mapfile "whitespace.map":

`EXPRESSION "Lake Superior " #no map generated`
`#EXPRESSION "Lake Superior"   #map is generated`

But this is exactly my issue. When the name-attribute of the feature in the database is 'Lake Superior ' (with one space at the end) then i would expect that the feature is drawn when i define `EXPRESSION "Lake Superior "` (with one space at the end) and not when i define `EXPRESSION "Lake Superior"` (without a space at the end) in the mapfile.

**Comment by sdlime** [comment]
I'll look at this tonight...

**Comment by jmckenna** [comment]
@glk-24 indeed all this confusion would have been prevented if you followed the ticket template/instructions when you filed the ticket, and spent the effort yourself in creating a small dump/mapfile/commands file for everyone to reproduce, and attached it here.  

I confirm that my test case archive does produce the PostGIS issue (it seems my image preview here locally was cached somehow).  I can trigger now if I generate different names for the map image requests.

**Comment by sdlime** [comment]
Initial report was fine IMHO, I was curious if it was limited to PostGIS or not since expression evaluation is handled independent of the underlying driver. Anyway, in PostGIS, the code that populates a shapeObj items is trimming trailing blank values, so they're gone in the shapeObj - hence the failed expression.

Removing the trim blanks call would fix this specific instance. However,  there are PostgreSQL data types that are blank padded - looks like fixed character(n) fields are (https://www.postgresql.org/docs/9.0/datatype-character.html) so a fix would have to consider the column type and trim blanks in certain cases. I don't know what's possible in that regard.

I did start a pull request with the trivial fix but that's not going to work. *sigh*

--Steve

**Comment by jmckenna** [comment]
Good find Steve.   (for others, the specific line: https://github.com/MapServer/MapServer/blob/e306cd034fa4d7b2c9e60d5ecb923e8693025e77/mappostgis.cpp#L2077

**Comment by jmckenna** [comment]
I personally always use TEXT data types for strings in PostgreSQL, and after re-reading the page your referenced I remember why, yikes.  Definitely tricky in how to handle this for MapServer (the 'character' and 'character(n)' data types, which are blank padded).

**Comment by sdlime** [comment]
Curious how Oracle (@msmitherdc) and SQL Server (@szekerest) drivers handle this.

**Comment by geographika** [comment]
There is no use of `msStringTrimBlanks` in the MSSQL driver. 
IMHO for fixed character length fields (which are padded with spaces) in a database it should be up to the user to trim any spaces in the DATA clause for use in expressions. 
DBFs are trickier as there is no handy trim function (AFAIK) you can add to a DATA clause, so maybe these should be trimmed still. 

Most client applications have to do the same for CHAR fields (so I nearly always use NVARCHAR, unless space is an issue) - see https://stackoverflow.com/a/176565/179520 for distinction.

**Comment by geographika** [comment]
Link to comment re the addition of msStringTrimBlanks: https://github.com/MapServer/MapServer/issues/2817#issuecomment-4939282

> 
> Author: pramsey
> Date: 2008/12/11 - 07:14
> This should be "fixed" as of f7355c2 (r8238), though I am not 100% comfortable with the fix at this moment. I am simply lopping off all the trailing spaces. This might be refined to only do so in cases where the field type is fixed width. Some experimentation will be needed to find the field types that exhibit the behavior of returning a fixed value for PQgetlength no matter how wide the actual data is.

**Comment by sdlime** [comment]
>DBFs are trickier as there is no handy trim function (AFAIK) you can add to a DATA clause, so maybe these should be trimmed still.

Yes, we'd leave that alone. There's no way to store strings with meaningful trailing zeros in a shapefile. Has never been an issue.

Removing the msStringTrimBlanks() call would be a regression (perhaps a significant one?), one that could be handled via documentation I suppose. Regardless, that's not a 7.6.x type of change. Refining would be a better option if there was a fast, reliable way to do so...
